src/js/png.js.rst

```diff
@@ -0,0 +1,90 @@
+PNG Handling
+------------
+Utilities for generating and reading PNG files.
+
+
+Convert a Layer to PNG
+~~~~~~~~~~~~~~~~~~~~~~
+Follows `Aby Nimbalkar`_'s `common.js`_ in Xport_.
+
+.. _Aby Nimbalkar: https://github.com/abynim
+.. _common.js: https://github.com/abynim/Xport/blob/master/Xport.sketchplugin/Contents/Sketch/common.js
+.. _Xport: https://github.com/abynim/Xport
+
+To export the layer to PNG, first create an MSExportRequest.
+
+.. code-block:: javascript
+
+    export function toFile(context, layer, scale) {
+
+      const slice = MSExportRequest.alloc().init();
+
+      slice.setName(layer.name);
+      slice.setFormat('png');
+      slice.setShouldTrim(0);
+      slice.setSaveForWeb(1);
+
+Use the dimensions of the input layer to set the export rectangle.
+
+.. code-block:: javascript
+
+      const rect = layer.sketchObject.absoluteRect().rect();
+
+      slice.setRect(rect);
+
+Specify an scaling if requested in the parameters.
+
+.. code-block:: javascript
+
+      if (typeof scale === 'undefined') {
+        scale = 1;
+      }
+
+      slice.setScale(scale);
+
+And perform the export.
+
+.. code-block:: javascript
+
+      const path = `${ NSTemporaryDirectory() }sketch.png`;
+      context.document.saveArtboardOrSlice_toFile(slice, path);
+
+      return path;
+    }
+
+
+Read a PNG File
+~~~~~~~~~~~~~~~
+To read a PNG file, we have to access the filesystem on the Cocoa side of the
+bridge which will give us an ``NSData`` object. Working with ``NSData`` is
+difficult. So we move JavaScript side by converting the ``NSData`` to a Base64
+string and traverse the bridge with this string.
+
+Once back on the JavaScript side we decode the Base64 data to a Buffer.
+
+.. code-block:: javascript
+
+    const base64Decoder = require('base64-arraybuffer').decode;
+    const arrayBufferToBuffer = require('arraybuffer-to-buffer');
+    const parsePng = require('pngparse-sync');
+
+    export function fromFile(path) {
+      const data = NSData.dataWithContentsOfFile(path);
+      const base64 = data.base64EncodedStringWithOptions(0);
+
+On the JavaScript side of the bridge this is still an MOBoxedObject so coerce it
+to a proper JavaScript string first.
+
+.. code-block:: javascript
+
+      const coerced = `${ base64 }`;
```

Comment:
Isn't `.toString()` more explicit?

Reply:
Uh... you're not going to like this...

`base64` was created on the Objective C side of the bridge. Over there it was a `NSString *` per https://developer.apple.com/documentation/foundation/nsdata/1413546-base64encodedstringwithoptions. I was expecting this to become a JavaScript String by conversation across the bridge but it's actually a MOBoxedObject instead. This is the boxing in the Mocha bridge for Objective C objects when they are JavaScript-side.

But the `toString` on `MOBoxedObject` gives `null`. And the string interpolation like here works.

I don't have an explanation for this.

The .toString doesn't error which means it can't have been translated to a call on the Objective C MOBoxedObject or NSString that it is boxing. (Because neither of these have a toString method defined on them and would bomb.) Which only leaves JavaScript. If the MOBoxedObject has a representation in the JavaScriptCore framework that this whole show is executing in then it might be that .toString is executing the Object prototype toString on a JavaScriptCore object corresponding to the MOBoxedObject? I don't understand CocoaScript enough to know and anyway it would explain the .toString null but not why the template string interpolation works.

Sorry, don't know.

Reply:
You're right. I don't like this. But thanks for the explanation, it makes sense - to some degree. I guess the alternative way to probably get to the same result would be `"" + base64` , but the string interpolation does indeed look nicer.

Reply:
Yeah, sorry. I hate this too.